sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Convert.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.beam.sdk.schemas.SchemaRegistry;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunctions;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/** A set of utilities for converting between different objects supporting schemas. */
+@Experimental(Kind.SCHEMAS)
+public class Convert {
+  /**
+   * Convert a {@link PCollection<InputT>} into a {@link PCollection<Row>}.
+   *
+   * <p>The input {@link PCollection} must have a schema attached. The output collection will have
+   * the same schema as the iput.
+   */
+  public static <InputT> PTransform<PCollection<InputT>, PCollection<Row>> toRows() {
+    return to(Row.class);
+  }
+
+  /**
+   * Convert a {@link PCollection<Row>} into a {@link PCollection<OutputT>}.
+   *
+   * <p>The output schema will be inferred using the schema registry. A schema must be registered
+   * for this type, or the conversion will fail.
+   */
+  public static <OutputT> PTransform<PCollection<Row>, PCollection<OutputT>> fromRows(
+      Class<OutputT> clazz) {
+    return to(clazz);
+  }
+
+  /**
+   * Convert a {@link PCollection<Row>} into a {@link PCollection<OutputT>}.
+   *
+   * <p>The output schema will be inferred using the schema registry. A schema must be registered
+   * for this type, or the conversion will fail.
+   */
+  public static <OutputT> PTransform<PCollection<Row>, PCollection<OutputT>> fromRows(
+      TypeDescriptor<OutputT> typeDescriptor) {
+    return to(typeDescriptor);
+  }
+
+  /**
+   * Convert a {@link PCollection<InputT>} to a {@link PCollection<OutputT>}.
+   *
+   * <p>This function allows converting between two types as long as the two types have
+   * <i>compatible</i> schemas. Two schemas are said to be <i>compatible</i> if they recursively
+   * have fields with the same names, but possibly different orders.
+   */
+  public static <InputT, OutputT> PTransform<PCollection<InputT>, PCollection<OutputT>> to(
+      Class<OutputT> clazz) {
+    return to(TypeDescriptor.of(clazz));
+  }
+
+  /**
+   * Convert a {@link PCollection<InputT>} to a {@link PCollection<OutputT>}.
+   *
+   * <p>This function allows converting between two types as long as the two types have
+   * <i>compatible</i> schemas. Two schemas are said to be <i>compatible</i> if they recursively
+   * have fields with the same names, but possibly different orders.
+   */
+  public static <InputT, OutputT> PTransform<PCollection<InputT>, PCollection<OutputT>> to(
+      TypeDescriptor<OutputT> typeDescriptor) {
+    return new ConvertTransform<>(typeDescriptor);
+  }
+
+  private static class ConvertTransform<InputT, OutputT>
+      extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+    @Nullable TypeDescriptor<OutputT> outputTypeDescriptor = null;
```

Comment:
Is this really `@Nullable`? Looks like it is always set in the constructor, you should leave off the default value and make it final.

Reply:
Fine to change for now. In the future, might have to make this nullable though.